msprime/mutations.py

```diff
@@ -24,57 +24,114 @@
-            self.__init__).parameters.keys() if hasattr(self, key)}
+    TODO document properly.
+    """
+    def __init__(self):
+        alleles = [b"0", b"1"]
+        root_distribution = [1, 0]
+        transition_matrix = [[0, 1], [1, 0]]
+        alleles = [b"0", b"1"]
+        super().__init__(alleles, root_distribution, transition_matrix)
 
 
-class MutationMap(object):
+class JukesCantor(MutationModel):
     """
-    TODO document
+    The Jukes-Cantor mutation model.
+
+    .. todo: documentation
     """
+    def __init__(self):
+        alleles = [b"A", b"C", b"T", b"G"]
+        root_distribution = [0.25, 0.25, 0.25, 0.25]
+        transition_matrix = np.zeros((4, 4))
+        transition_matrix[:] = 1 / 3
+        np.fill_diagonal(transition_matrix, 0)
+        super().__init__(alleles, root_distribution, transition_matrix)
+
+
+# Pre 1.0, we had these constants to define the alphabet for the
+# simple infinite sites mutations. These should be deprecated and removed
+# along with the InfiniteSites class.
+BINARY = 0
+NUCLEOTIDES = 1
+
+
+class InfiniteSites(MutationModel):
+    # This mutation model is defined for backwards compatability, and is a remnant
+    # avoid confusion. The class should be formally deprecated and removed at
+    # some point.
+    def __init__(self, alphabet=BINARY):
+        self.alphabet = alphabet
+        models = {
+            BINARY: Binary(),
+            NUCLEOTIDES: JukesCantor()
+        }
+        if alphabet not in models:
+            raise ValueError("Bad alphabet")
+        model = models[alphabet]
+        super().__init__(model.alleles, model.root_distribution, model.transition_matrix)
+
+
+class MutationMap(object):
+    # TODO Get rid of this class and use IntervalMap (or maybe IntervalRateMap) instead.
+    # See
+    # https://github.com/tskit-dev/msprime/issues/902
+    # https://github.com/tskit-dev/msprime/issues/920
     def __init__(self, position, rate):
         self.position = position
         self.rate = rate
+        self._ll_map = _msprime.IntervalMap(position=position, value=rate)
+
+    def asdict(self):
+        return {"position": self.position, "rate": self.rate}
+
+
+def _simple_mutation_generator(rate, sequence_length, rng):
+    """
+    Factory function used to create a low-level mutation generator that
+    produces binary infinite sites mutations suitable for use in
+    msprime.simulate() and the mspms CLI.
+    """
+    # Add a ``discrete`` parameter and pass to the MutationGenerator
+    # constructor.
+    if rate is None:
+        return None
+    rate_map = MutationMap(
+        position=[0, sequence_length],
+        rate=[rate, 0]
+    )
+    return _msprime.MutationGenerator(rng, rate_map._ll_map, Binary())
 
 
 def mutate(
-        tree_sequence, rate=None, random_seed=None, model=None, keep=False,
-        start_time=None, end_time=None):
+        tree_sequence, rate=0, random_seed=None, model=None, keep=False,
```

Comment:
Why change the default for rate to zero here @petrelharp? I think it's much better to have None as the default unless you are absolutely certain that the default behaviour in the context of other parameters will never change. We don't know what future parameters will be, and it might turn on that the sensible default behaviour then is to have a non-zero value.

Reply:
This is also a breaking change - people could be using None as an input right now.

Reply:
My bad!